apps/emqx/test/emqx_session_mem_SUITE.erl

```diff
@@ -0,0 +1,613 @@
+%%--------------------------------------------------------------------
+
+t_subscribe(_) ->
+    ok = meck:expect(emqx_broker, subscribe, fun(_, _, _) -> ok end),
+    {ok, Session} = emqx_session_mem:subscribe(<<"#">>, subopts(), session()),
+    ?assertEqual(1, emqx_session_mem:info(subscriptions_cnt, Session)).
+
+t_is_subscriptions_full_false(_) ->
+    Session = session(#{max_subscriptions => infinity}),
+    ?assertNot(emqx_session_mem:is_subscriptions_full(Session)).
+
+t_is_subscriptions_full_true(_) ->
+    ok = meck:expect(emqx_broker, subscribe, fun(_, _, _) -> ok end),
+    Session = session(#{max_subscriptions => 1}),
+    ?assertNot(emqx_session_mem:is_subscriptions_full(Session)),
+    {ok, Session1} = emqx_session_mem:subscribe(
+        <<"t1">>, subopts(), Session
+    ),
+    ?assert(emqx_session_mem:is_subscriptions_full(Session1)),
+    {error, ?RC_QUOTA_EXCEEDED} = emqx_session_mem:subscribe(
+        <<"t2">>, subopts(), Session1
+    ).
+
+t_unsubscribe(_) ->
+    ok = meck:expect(emqx_broker, unsubscribe, fun(_) -> ok end),
+    SubOpts = subopts(),
+    Session = session(#{subscriptions => #{<<"#">> => SubOpts}}),
+    {ok, Session1, SubOpts} = emqx_session_mem:unsubscribe(<<"#">>, Session),
+    {error, ?RC_NO_SUBSCRIPTION_EXISTED} = emqx_session_mem:unsubscribe(<<"#">>, Session1).
+
+t_publish_qos0(_) ->
+    ok = meck:expect(emqx_broker, publish, fun(_) -> [] end),
+    Msg = emqx_message:make(clientid, ?QOS_0, <<"t">>, <<"payload">>),
+    {ok, [], [], Session} = emqx_session_mem:publish(1, Msg, Session = session()),
+    {ok, [], [], Session} = emqx_session_mem:publish(undefined, Msg, Session).
+
+t_publish_qos1(_) ->
+    ok = meck:expect(emqx_broker, publish, fun(_) -> [] end),
+    Msg = emqx_message:make(clientid, ?QOS_1, <<"t">>, <<"payload">>),
+    {ok, [], [], Session} = emqx_session_mem:publish(1, Msg, Session = session()),
+    {ok, [], [], Session} = emqx_session_mem:publish(2, Msg, Session).
+
+t_publish_qos2(_) ->
+    ok = meck:expect(emqx_broker, publish, fun(_) -> [] end),
+    Msg = emqx_message:make(clientid, ?QOS_2, <<"t">>, <<"payload">>),
+    {ok, [], Session} = emqx_session_mem:publish(1, Msg, session()),
+    ?assertTimerSet(expire_awaiting_rel, _Timeout),
+    ?assertEqual(1, emqx_session_mem:info(awaiting_rel_cnt, Session)),
+    {ok, Session1} = emqx_session_mem:pubrel(1, Session),
+    ?assertEqual(0, emqx_session_mem:info(awaiting_rel_cnt, Session1)),
+    {error, ?RC_PACKET_IDENTIFIER_NOT_FOUND} = emqx_session_mem:pubrel(1, Session1).
+
+t_publish_qos2_with_error_return(_) ->
+    ok = meck:expect(emqx_broker, publish, fun(_) -> [] end),
+    ok = meck:expect(emqx_hooks, run, fun
+        ('message.dropped', [Msg, _By, ReasonName]) ->
+            self() ! {'message.dropped', ReasonName, Msg},
+            ok;
+        (_Hook, _Arg) ->
+            ok
+    end),
+
+    Session = session(#{max_awaiting_rel => 2, awaiting_rel => #{PacketId1 = 1 => ts(millisecond)}}),
+    begin
```

Comment:
Unrelated to the PR, but `begin`/`end` block does absolutely nothing here. In the future it can be replaced with comment: `Ensure that publishing a new message with the packet ID that matches ID of a packet awaiting release generates an exception.`